src/main/java/org/jitsi/videobridge/Content.java

```diff
@@ -239,53 +245,91 @@ public RtpChannel createRtpChannel(String channelBundleId,
                                        RTPLevelRelayType rtpLevelRelayType)
         throws Exception
     {
-        RtpChannel channel = null;
+        return createRtpChannel(
+                channelBundleId, transportNamespace,
+                initiator, rtpLevelRelayType,
+                false /* octo */);
+    }
+
+    /**
+     * Initializes a new <tt>RtpChannel</tt> instance and adds it to the list of
+     * <tt>RtpChannel</tt>s of this <tt>Content</tt>. The new
+     * <tt>RtpChannel</tt> instance has an ID which is unique within the list of
+     * <tt>RtpChannel</tt>s of this <tt>Content</tt>.
+     *
+     * @param channelBundleId the ID of the channel-bundle that the created
+     * <tt>RtpChannel</tt> is to be a part of (or <tt>null</tt> if it is not to
+     * be a part of a channel-bundle).
+     * @param transportNamespace transport namespace that will used by new
+     * channel. Can be either {@link IceUdpTransportPacketExtension#NAMESPACE}
+     * or {@link RawUdpTransportPacketExtension#NAMESPACE}.
+     * @param initiator the value to use for the initiator field, or
+     * <tt>null</tt> to use the default value.
+     * @param rtpLevelRelayType
+     * @return the created <tt>RtpChannel</tt> instance.
+     * @param octo whether to create a regular channel or an Octo channel.
+     * @throws Exception
+     */
+    public RtpChannel createRtpChannel(String channelBundleId,
+                                       String transportNamespace,
+                                       Boolean initiator,
+                                       RTPLevelRelayType rtpLevelRelayType,
+                                       boolean octo)
+        throws Exception
+    {
+        RtpChannel channel;
 
-        do
+        synchronized (channels)
```

Comment:
does this synchronized block need to be this big?  doesn't it just need to protect the `channels.put(id, channel);` ?  i see that `generateUniqueChannelID` already locks it on its own.

Reply:
We do need it. If we release the lock before the call to channels.put() it is possible for another thread to call generateUniqueChannelID() and receive the same ID.

The block has many lines of code, but the only thing it does (apart from generating an ID and channels.put()) is to instantiate a channel. If readability is a concern I can extract that logic to a separate method, but as for reducing the code that executes inside the synchronized block, I don't think there's anything to do.

Reply:
not sure i follow, why not just remove it from here and change `channels.put(id, channel);` to:
```
synchronized (channels) {
  channels.put(id, channel);
}
```

Reply:
Thread1 calls generateUniqueChannelID(), gets "a" and yields.
Thread2 calls generateUniqueChannelID() and also gets "a" (this is possible because Thread1 yielded before adding an entry for "a").

Both threads instantiate a channel with ID "a".

Reply:
oops, missed that.  nevermind then :)